### ISSUE TITLE
Sudden performance degradation during the training stage !!!!

### ISSUE THREAD

**Issue by XiaokunFeng** [lizhou-cs/JointNLT#15]
Hello, thanks for your great job!
I followed the steps provided to train the model from scratch (swin_b_ep300.yaml, with 4 3090 GPU).  But I found that the model suddenly experienced a sudden drop in performance after 70 epochs of training (as shown in the figure below), and before the drop, the model's iou (sum) performance could reach 2.4 on the training set.

I tried to continue training a batch based on the provided model (JointNLT_ep0300.pth.tar), it was found that the model's iou (sum) performance can reach around 2.7 on the training set.

So, the model was not fully trained in 70 epochs. I'm curious why there was a sudden drop in performance. May I ask if there are any details that need to be noted? (I noticed that the name of provided model (JointNLT_ep0300.pth.tar) is called TransNLT, but the model' name in the open source code is called JointNLT)

Looking forward to your reply. Thanks again for your great work.
<img width="926" alt="image" src="https://github.com/lizhou-cs/JointNLT/assets/103614531/e8b2d8c1-3e80-48ec-a65f-7397d49ab8a2">

**Comment by jaypatravali** [comment]
Another issue i noted was that L191 in https://github.com/lizhou-cs/JointNLT/blob/94b338e19f7bc4a4d9646b0f0ac82321e9bddb26/lib/train/trainers/base_trainer.py#L191 

checkpoint_dict = torch.load(checkpoint_path, map_location='cpu') 

checkpoint_dict[key] is 200 in the JointNLT_OTB_NL.pth.tar and JointNLT_ep0300.pth.tar. Shouldnt it be 300?

**Comment by XiaokunFeng** [comment]
I also noticed this situation (checkpoint_dict[key] is 200). May I ask if you have also met a sudden drop in performance?

**Comment by jaypatravali** [comment]
@XiaokunFeng Do you mean a sudden drop in training performance using these weights for finetuning?

I used  JointNLT_ep0300.pth.tar and use analyze_results.py on OTB dataset and receive 

Computed results over 48 / 48 sequences
otb           | AUC        | OP50       | OP75       | Precision    | Norm Precision    |
Transnlt      | 57.55      | 71.89      | 36.46      | 75.59        | 68.43             |

Using JointNLT_OTB_NL.pth.tar I received these results.

Reporting results over 48 / 48 sequences
otb           | AUC        | OP50       | OP75       | Precision    | Norm Precision    |
Transnlt      | 58.13      | 73.28      | 36.86      | 75.80        | 69.27             |

@lizhou-cs can you please assist?

**Comment by XiaokunFeng** [comment]
Sorry for my vague expressions.
I train the model from scratch (without using the provided model weight). As shown in the above fig, the model is trained from epoch = 0. And in this process, I found the sudden drop.

The reason why I use the the provided model (JointNLT_ep0300.pth.tar) to continue training ( just a batch)  is to checkout what the trained model's performance in training set.

Besides, I also continued to train the provided model (JointNLT_ep0300.pth.tar) for more 100 epochs in the past two days. I didn't find the sudden drop.

Could you please use the provided code (swin_b_ep300.yaml ) to train the model from scratch? Thanks!

**Comment by lizhou-cs** [comment]
> Hello, thanks for your great job! I followed the steps provided to train the model from scratch (swin_b_ep300.yaml, with 4 3090 GPU). But I found that the model suddenly experienced a sudden drop in performance after 70 epochs of training (as shown in the figure below), and before the drop, the model's iou (sum) performance could reach 2.4 on the training set.
> 
> I tried to continue training a batch based on the provided model (JointNLT_ep0300.pth.tar), it was found that the model's iou (sum) performance can reach around 2.7 on the training set.
> 
> So, the model was not fully trained in 70 epochs. I'm curious why there was a sudden drop in performance. May I ask if there are any details that need to be noted? (I noticed that the name of provided model (JointNLT_ep0300.pth.tar) is called TransNLT, but the model' name in the open source code is called JointNLT)
> 
> Looking forward to your reply. Thanks again for your great work. <img alt="image" width="926" src="https://user-images.githubusercontent.com/103614531/265178760-e8b2d8c1-3e80-48ec-a65f-7397d49ab8a2.png">


Hello, thank you for your interest in our work. I haven't encountered this issue before. Could you clarify if this problem consistently occurs after the 70th epoch? If it does, you might consider visualizing the images in that batch, as it could be caused by anomalous data. To provide more accurate advice, I would need additional details.

**Comment by lizhou-cs** [comment]
> Another issue i noted was that L191 in
> 
> https://github.com/lizhou-cs/JointNLT/blob/94b338e19f7bc4a4d9646b0f0ac82321e9bddb26/lib/train/trainers/base_trainer.py#L191
> 
> checkpoint_dict = torch.load(checkpoint_path, map_location='cpu')
> 
> checkpoint_dict[key] is 200 in the JointNLT_OTB_NL.pth.tar and JointNLT_ep0300.pth.tar. Shouldnt it be 300?

Hi, I replied to this issue in this [link](https://github.com/lizhou-cs/JointNLT/issues/11).

**Comment by lizhou-cs** [comment]
> XiaokunFeng

Hello, based on my previous experimental experience, the evaluation results on the OTB dataset may not accurately reflect the tracking performance of this model. This is primarily because the OTB dataset is relatively small, and its labeling standards are not consistent with other datasets, which can potentially lead to overfitting issues. Therefore, if you intend to fine-tune the model, it's advisable to choose LaSOT or TNL2K for evaluation, as these two datasets can better reflect the model's convergence.

**Comment by XiaokunFeng** [comment]
> > Hello, thanks for your great job! I followed the steps provided to train the model from scratch (swin_b_ep300.yaml, with 4 3090 GPU). But I found that the model suddenly experienced a sudden drop in performance after 70 epochs of training (as shown in the figure below), and before the drop, the model's iou (sum) performance could reach 2.4 on the training set.
> > I tried to continue training a batch based on the provided model (JointNLT_ep0300.pth.tar), it was found that the model's iou (sum) performance can reach around 2.7 on the training set.
> > So, the model was not fully trained in 70 epochs. I'm curious why there was a sudden drop in performance. May I ask if there are any details that need to be noted? (I noticed that the name of provided model (JointNLT_ep0300.pth.tar) is called TransNLT, but the model' name in the open source code is called JointNLT)
> > Looking forward to your reply. Thanks again for your great work. <img alt="image" width="926" src="https://user-images.githubusercontent.com/103614531/265178760-e8b2d8c1-3e80-48ec-a65f-7397d49ab8a2.png">
> 
> Hello, thank you for your interest in our work. I haven't encountered this issue before. Could you clarify if this problem consistently occurs after the 70th epoch? If it does, you might consider visualizing the images in that batch, as it could be caused by anomalous data. To provide more accurate advice, I would need additional details.

@lizhou-cs 
Thanks for your advice. I carefully analyse my experiment results. I have a few questions to discuss with you.

Firstly, I don't think the sudden drop in performance is caused by anomalous data. The reasons are as follows: 
1. I have trained other models with those datasets before, and I didn't find a sudden drop in performance.
2. When I continued to train based on  your provided model weights (train for 150 epoches), I did not find sudden drop; But here, after 70 epoches of training from scratch, there was a sudden drop in performance. (Because training data are randomly sampled, if it is due to anomalous data, 150 rounds of training are more likely to be affected by anomalous data)

Besides, I rechek the log file, and I found a strange phenomenon. As for the total loss (sum of g_loss, t1_loss and t2_loss), I plotted the relevant performance indicators around the sudden drop in performance, as shown in the following figure :
![image](https://github.com/lizhou-cs/JointNLT/assets/103614531/567d54f7-a202-46a3-b857-cd541a043301)
We can see that after a sudden performance drop, the loss (giou, l1 and toyal) increases, while the iou value decreases.

Furthermore, I plotted the performance indicators of g_loss, t1_loss, t2_loss around the sudden drop in performance, as shown in the following figure :
![image](https://github.com/lizhou-cs/JointNLT/assets/103614531/5de2d7f5-295a-4fc9-95b8-2ebd29ab2e00)

![image](https://github.com/lizhou-cs/JointNLT/assets/103614531/b2862575-3a67-44e5-ac42-2d431b679255)

![image](https://github.com/lizhou-cs/JointNLT/assets/103614531/24957a19-178e-43a0-b2fe-aaf87854ba9a)

As can be seen, the changes in the four indicators of g_loss are similar to those of loss ( the loss (giou, l1 and toyal) increases, while the iou value decreases).

Strangely, the changes in the four indicators of l1_loss and l2_loss are different. the loss (giou, l1 and toyal)  still decreases, but the iou value decreases too!

From the above experimental phenomena, it can be concluded that the model has indeed reduced l2_ Loss and l2_loss after training. But the decrease in loss leads to a decrease in the actual performance (iou) of the model?

Can you help explain the above phenomenon? Thank you!

**Comment by XiaokunFeng** [comment]
> > > Hello, thanks for your great job! I followed the steps provided to train the model from scratch (swin_b_ep300.yaml, with 4 3090 GPU). But I found that the model suddenly experienced a sudden drop in performance after 70 epochs of training (as shown in the figure below), and before the drop, the model's iou (sum) performance could reach 2.4 on the training set.
> > > I tried to continue training a batch based on the provided model (JointNLT_ep0300.pth.tar), it was found that the model's iou (sum) performance can reach around 2.7 on the training set.
> > > So, the model was not fully trained in 70 epochs. I'm curious why there was a sudden drop in performance. May I ask if there are any details that need to be noted? (I noticed that the name of provided model (JointNLT_ep0300.pth.tar) is called TransNLT, but the model' name in the open source code is called JointNLT)
> > > Looking forward to your reply. Thanks again for your great work. <img alt="image" width="926" src="https://user-images.githubusercontent.com/103614531/265178760-e8b2d8c1-3e80-48ec-a65f-7397d49ab8a2.png">
> > 
> > 
> > Hello, thank you for your interest in our work. I haven't encountered this issue before. Could you clarify if this problem consistently occurs after the 70th epoch? If it does, you might consider visualizing the images in that batch, as it could be caused by anomalous data. To provide more accurate advice, I would need additional details.
> 
> @lizhou-cs Thanks for your advice. I carefully analyse my experiment results. I have a few questions to discuss with you.
> 
> Firstly, I don't think the sudden drop in performance is caused by anomalous data. The reasons are as follows:
> 
> 1. I have trained other models with those datasets before, and I didn't find a sudden drop in performance.
> 2. When I continued to train based on  your provided model weights (train for 150 epoches), I did not find sudden drop; But here, after 70 epoches of training from scratch, there was a sudden drop in performance. (Because training data are randomly sampled, if it is due to anomalous data, 150 rounds of training are more likely to be affected by anomalous data)
> 
> Besides, I rechek the log file, and I found a strange phenomenon. As for the total loss (sum of g_loss, t1_loss and t2_loss), I plotted the relevant performance indicators around the sudden drop in performance, as shown in the following figure : ![image](https://user-images.githubusercontent.com/103614531/266838001-567d54f7-a202-46a3-b857-cd541a043301.png) We can see that after a sudden performance drop, the loss (giou, l1 and toyal) increases, while the iou value decreases.
> 
> Furthermore, I plotted the performance indicators of g_loss, t1_loss, t2_loss around the sudden drop in performance, as shown in the following figure : ![image](https://user-images.githubusercontent.com/103614531/266838071-5de2d7f5-295a-4fc9-95b8-2ebd29ab2e00.png)
> 
> ![image](https://user-images.githubusercontent.com/103614531/266838099-b2862575-3a67-44e5-ac42-2d431b679255.png)
> 
> ![image](https://user-images.githubusercontent.com/103614531/266838115-24957a19-178e-43a0-b2fe-aaf87854ba9a.png)
> 
> As can be seen, the changes in the four indicators of g_loss are similar to those of loss ( the loss (giou, l1 and toyal) increases, while the iou value decreases).
> 
> Strangely, the changes in the four indicators of l1_loss and l2_loss are different. the loss (giou, l1 and toyal) still decreases, but the iou value decreases too!
> 
> From the above experimental phenomena, it can be concluded that the model has indeed reduced l2_ Loss and l2_loss after training. But the decrease in loss leads to a decrease in the actual performance (iou) of the model?
> 
> Can you help explain the above phenomenon? Thank you!

This is the model weight and yaml file saved before a sudden performance drop occurs (link: https://pan.baidu.com/s/1f3BbD1flexTLp2IHhchh1g passward: gxhd 
). Based on this model weight , I continued to train twice and found that the model experienced a sudden performance drop after training 1 and 2 epoches, respectively.

I would be very grateful and it would be of great help to me, if you could try to continue training based on this model weight, and analyze the potential performance drop issues that may occur.

**Comment by xwhzz** [comment]
I also meet this situation at the 65th epoch.

**Comment by lwenzh** [comment]
Has the problem been resolved? When I train the model from scratch, I also meet this situation around the 30th epoch.

**Comment by Jilrock-c** [comment]
Have you solved this problem? I encountered it at the 19th epoch.

**Comment by ssaishandsome** [comment]
同样遇到了此问题

**Comment by XiaokunFeng** [comment]
I tried reducing the learning rate to 0.1 of the original value and found that there was no sudden drop in performance. However, the performance of the final trained model is weaker compared to the performance shown in the paper.

**Comment by ssaishandsome** [comment]
> 我尝试将学习率降低到原始值的 0.1，发现性能并没有突然下降。但是，与论文中所示的性能相比，最终训练模型的性能要弱一些。

能告知我您能达到什么样的性能吗？

**Comment by Jilrock-c** [comment]
> I tried reducing the learning rate to 0.1 of the original value and found that there was no sudden drop in performance. However, the performance of the final trained model is weaker compared to the performance shown in the paper.

Thank you very much for your tips. I would like to confirm whether the learning rate of the version you did not encounter this problem is 0.0002
